chips/stm32f4xx/src/flash.rs

```diff
@@ -0,0 +1,549 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright OxidOS Automotive SRL.
+
+#![deny(dead_code)]
+#![deny(missing_docs)]
+#![deny(unused_imports)]
+//! STM32F4xx flash driver
+//!
+//! This driver provides basic functionalities for the entire STM32F4 series.
+//!
+//! # Features
+//!
+//! - [x] Configuring latency based on the system clock frequency
+//!
+//! # Missing features
+//!
+//! - [ ] Support for different power supplies
+//! - [ ] Instruction prefetch
+//! - [ ] Instruction and data cache
+//!
+//!
+//! # Usage
+//!
+//! To use this driver, a reference to the Flash peripheral is required:
+//!
+//! ```rust,ignore
+//! // Inside the board main.rs
+//! let flash = &peripherals.stm32f4.flash;
+//! ```
+//!
+//! ## Retrieve the current flash latency
+//!
+//! ```rust,ignore
+//! let flash_latency = flash.get_latency() as usize;
+//! debug!("Current flash latency is {}", flash_latency);
+//! ```
+
+use crate::chip_specific::flash_specific::FlashLatency;
+use crate::chip_specific::flash_specific::SpecificFlashTrait;
+
+use kernel::debug;
+use kernel::utilities::registers::interfaces::{ReadWriteable, Readable};
+use kernel::utilities::registers::{register_bitfields, register_structs, ReadWrite, WriteOnly};
+use kernel::utilities::StaticRef;
+use kernel::ErrorCode;
+
+register_structs! {
+    /// FLASH
+    FlashRegisters {
+        /// Flash access control register
+        (0x000 => acr: ReadWrite<u32, ACR::Register>),
+        /// Flash key register
+        (0x004 => keyr: WriteOnly<u32>),
+        /// Flash option key register
+        (0x008 => optkeyr: WriteOnly<u32>),
+        /// Status register
+        (0x00C => sr: ReadWrite<u32, SR::Register>),
+        /// Control register
+        (0x010 => cr: ReadWrite<u32, CR::Register>),
+        /// Flash option control register
+        (0x014 => optcr: ReadWrite<u32, OPTCR::Register>),
+        /// Flash option control register 1
+        // NOTE: This register is present only on some chip models
+        (0x018 => optcr1: ReadWrite<u32>),
```

Comment:
This is one of the cases where we explicitly allow use of `#cfg[` based on the specific chip in use. As example, `chips/nrf5x/*` uses the pattern somewhat heavily to differentiate the hardware physically available on the `nrf51` vs `nrf52` family. It should be compile-time impossible to try to access a register on a chip that doesn't have it.

Reply:
Though I fear I'm slightly contradicting or giving confusing feedback with @bradjc's comment here: https://github.com/tock/tock/pull/3523#pullrequestreview-1503774449

I think the balance we've come to to-date is that generic `abcXX` crate should use `cfg` to define the presence / absence of hardware correctly, but any implementation or function/etc that would try to _use_ that register should not be `cfg`'d around in the `abcXX` crate, but rather should be in the specific `abc22` chip crate that actually has the hardware.

Reply:
I wasn't able to achieve this using `register_structs!` macro. Any help?

Reply:
Ohh... interesting... nrf5x gpio, temperature, etc don't use the `register_structs!` macro — they have hand-written register structs (likely because they predate that macro...). Frankly, it is _hard_ to read the `cfg`'d struct with the mixed chip definitions as they get bigger... it looks like your latest code here just went with the 'hand-written register struct' approach. I think that is a fine stopgap, especially for a small device like this, but long-term this is probably something that `register_structs!` should figure out how to support (at risk of making that macro even more complex... :/ )